scripts/core/editor3/actions/spellchecker.tsx

```diff
@@ -36,9 +37,10 @@ export function reloadSpellcheckerWarnings() {
     };
 }
 
-export function setSpellcheckerLanguage() {
+export function setSpellcheckerLanguage(language: string) {
```

Comment:
Why take language as a parameter? Is there a case when some other language than currently selected is used?

Reply:
Sometimes when we frequently change the language of an item, `state.spellchecking` holds the previous language so to make sure it holds the current language only passing it as a parameter.

Reply:
Have you tried to debug what is happening? We should try to fix the incorrect state rather than working around it. If you can't figure it out, describe me how to reproduce it and I'll take a look.

Reply:
I did, and both the `state.item.language` and `state.spellchecking.language` not getting updated immediately. 
How to reproduce:
Remove language as a parameter from here `export function setSpellcheckerLanguage(language: string)`
It happens only when you type something in the body or headline, change the language and then undo the changes using `ctrl+z` and then again change the language, spellchecker would run with the previous language, not with the current one.

Reply:
I've been checking out why it behaves that way and found that not only editor3 store changes, but also the article item itself which shouldn't be related to editor3. Can you debug what is causing the change? here's a video showing `item.language` change: https://drive.google.com/open?id=1qAmcvm3TcoJUQTFTgx3E4v9IKpfNX_WZ

Reply:
Do you mean undoing changes shouldn't impact header fields?

Reply:
No. If you are editing a slugline for example and hit ctrl+z, it's all fine that it undoes the change. The issue is that I was focused on editor3 field in the content section when hitting ctrl+z. If I undo while being focused on one field, it shouldn't undo a different field. Did you find the code that does it?

Reply:
I checked it, and what I found is that we are watching and storing the whole `item` in the history stack, whenever we make a change to any field of the item a new version of that item is pushed into the history stack. So, when we click `ctrl+z`, the previous version of the item from the history stack is applied to the `scope.item`. If we undo once or twice our focus remains on the current field that we are editing depending on how much we have changed on that field. But when we keep on pressing `ctrl+z` continuously it will apply the last version of item from the history stack i.e. version of an item when we last clicked on `save` button.
That means undo operation has nothing to do with the field being focused or not, it depends how many times we click on `ctrl+z` and how many versions are available in the history stack.

Reply:
Thanks for checking it. That's interesting. I never knew the feature existed. I've created SDESK-4569 to fix undo on the field causing global history to undo.

Anyway, spellchecker should work when global undo is used when no fields are focused. It seems to me that the underlying reason you are working around by adding a language parameter is that when you undo a global history, first the spellchecker actions are dispatched and only then the item is changed by the code that manages global history stack. That's probably why you have old data in state. Can you look into this and try to ensure that spellchecker actions are dispatched **after** the item history updated?

Reply:
I will look into it. And btw we are using this [angular history](https://github.com/decipherinc/angular-history) for managing this undo and redo.

Reply:
I looked at it but couldn't find any way to solve this issue. While investigating I found that history service broadcasts [this](https://github.com/superdesk/superdesk-client-core/blob/master/scripts/apps/authoring/authoring/directives/ArticleEditDirective.ts#L113) message and then we apply the changes. 

Also what to do if external spellcheckers are not found in the backend? We should show some message to the user or simply return an empty array or maybe both?

Reply:
> when you undo a global history, first the spellchecker actions are dispatched and only then the item is changed by the code that manages global history stack

Have you checked if this assumption is correct?

> Also what to do if external spellcheckers are not found in the backend? We should show some message to the user or simply return an empty array or maybe both?

If spellchecker is not found, display an error message and disable spellchecker for that article, without changing user preferences. If that's hard to do, return an empty array.

Reply:
> Have you checked if this assumption is correct?

Spellchecker actions are dispatched only after item gets changed as we are watching the `item.language` and if languages changes then only we run the spellchecker. But the state and store are not updated.

Reply:
Let's keep the language parameter then, but add a **clear** comment describing this issue.

Reply:
For now, I am returning an empty array, as we don't have any way to know if external spellchecker is available or not. If I try to show a message inside the check method it will keep popping up every time we type something. So, not sure but maybe we can create an endpoint to get external spellchecker status based on language.